lib/seattleflu/db/cli/command/etl/longitudinal.py

```diff
@@ -0,0 +1,356 @@
+from typing import Any
+from seattleflu.db.session import DatabaseSession
+from seattleflu.db.datatypes import Json
+from . import etl, find_or_create_site, upsert_individual, upsert_encounter
+from . import sample_identifier, find_sample, update_sample
+from . import UnknownSiteError, UnknownRaceError, UnknownEthnicGroupError
+from . import UnknownFluShotResponseError, UnknownInsuranceResponseError
+from .presence_absence import SampleNotFoundError
+
+
+LOG = logging.getLogger(__name__)
+
+
+# This revision number is stored in the processing_log of each longitudinal
+# record when the longitudinal record is successfully processed by this ETL
+# routine. The routine finds new-to-it records to process by looking for
+# longitudinal records lacking this revision number in their log.  If a
+# change to the ETL routine necessitates re-processing all longitudinal records,
+# this revision number should be incremented.
+REVISION = 1
+
+
+@etl.command("longitudinal", help = __doc__)
+
+@click.option("--dry-run", "action",
+    help        = "Only go through the motions of changing the database (default)",
+    flag_value  = "rollback",
+    default     = True)
+
+@click.option("--prompt", "action",
+    help        = "Ask if changes to the database should be saved",
+    flag_value  = "prompt")
+
+@click.option("--commit", "action",
+    help        = "Save changes to the database",
+    flag_value  = "commit")
+
+def etl_longitudinal(*, action: str):
+    LOG.debug(f"Starting the longitudinal ETL routine, revision {REVISION}")
+
+    db = DatabaseSession()
+
+    # Fetch and iterate over longitudinal records that aren't processed
+    #
+    # Rows we fetch are locked for update so that two instances of this
+    # command don't try to process the same longitudinal records.
+    LOG.debug("Fetching unprocessed longitudinal records")
+
+    longitudinal = db.cursor("longitudinal")
+    longitudinal.execute("""
+        select longitudinal_id as id, document
+          from receiving.longitudinal
+         where not processing_log @> %s
+         order by id
+           for update
+        """, (Json([{ "revision": REVISION }]),))
+
+    processed_without_error = None
+
+    try:
+        for record in longitudinal:
+            with db.savepoint(f"longitudinal record {record.id}"):
+                LOG.info(f"Processing longitudinal record {record.id}")
+
+                # Check validity of barcode
+                received_sample_identifier = sample_identifier(db,
+                    record.document["barcode"])
+
+                # Skip row if no matching identifier found
+                if received_sample_identifier is None:
+                    LOG.info("Skipping due to unknown barcode " + \
+                              f"{record.document['barcode']}")
+                    mark_skipped(db, record.id)
+                    continue
+
+                # Check sample exists in database
+                sample = find_sample(db,
```

Comment:
I'm a little confused by the order of how longitudinal data is added? When do we expect the samples for these to be created in the database?

Reply:
Good question. I don't have a good grasp here of the process.

Reply:
Yep, I think these samples end up being processed by Helen's lab and recorded in the manifest as normal.